// ask-smapi-sdk/dist/smapiClientBuilder/AbstractSmapiClientBuilder.js
"use strict";
/*
 * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.SmapiClientBuilder = void 0;
/**
 * Abstract Builder class which should be implemented.
 * @export
 * @class SmapiClientBuilder
 */
var SmapiClientBuilder = /** @class */ (function () {
    function SmapiClientBuilder() {
    }
    SmapiClientBuilder.prototype.withCustomUserAgent = function (userAgent) {
        this.customUserAgent = userAgent;
        return this;
    };
    SmapiClientBuilder.prototype.withRefreshTokenConfig = function (refreshTokenConfig) {
        this.refreshTokenConfig = refreshTokenConfig;
        return this;
    };
    SmapiClientBuilder.prototype.client = function () {
        throw new Error('client funtion is not implemented');
    };
    return SmapiClientBuilder;
}());
exports.SmapiClientBuilder = SmapiClientBuilder;
