// ask-smapi-sdk/dist/smapiClientBuilder/SmapiClientBuilder.js
"use strict";
/*
 * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.CustomSmapiClientBuilder = exports.StandardSmapiClientBuilder = void 0;
var ask_smapi_model_1 = require("ask-smapi-model");
var DefaultApiClient = ask_smapi_model_1.runtime.DefaultApiClient;
var AbstractSmapiClientBuilder_1 = require("./AbstractSmapiClientBuilder");
var DEFAULT_API_ENDPOINT = 'https://api.amazonalexa.com';
/**
 * StandardSmapiClientBuilder class use default ApiClient and default ApiEndpoint
 */
var StandardSmapiClientBuilder = /** @class */ (function (_super) {
    __extends(StandardSmapiClientBuilder, _super);
    function StandardSmapiClientBuilder() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * Funtion used to generate SkillManagementService instance.
     */
    StandardSmapiClientBuilder.prototype.client = function () {
        if (this.refreshTokenConfig) {
            var apiConfiguration = {
                apiClient: new DefaultApiClient(),
                apiEndpoint: DEFAULT_API_ENDPOINT,
                authorizationValue: null,
            };
            var authenticationConfiguration = {
                clientId: this.refreshTokenConfig.clientId,
                clientSecret: this.refreshTokenConfig.clientSecret,
                refreshToken: this.refreshTokenConfig.refreshToken,
            };
            return new ask_smapi_model_1.services.skillManagement.SkillManagementServiceClient(apiConfiguration, authenticationConfiguration, this.customUserAgent);
        }
        throw new Error('Please provide refreshToken Config to build smapi client');
    };
    return StandardSmapiClientBuilder;
}(AbstractSmapiClientBuilder_1.SmapiClientBuilder));
exports.StandardSmapiClientBuilder = StandardSmapiClientBuilder;
/**
 * CustomSmapiClientBuilder give user ability to configure Apiclient and ApiEndpoint
 */
var CustomSmapiClientBuilder = /** @class */ (function (_super) {
    __extends(CustomSmapiClientBuilder, _super);
    function CustomSmapiClientBuilder() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    CustomSmapiClientBuilder.prototype.withApiEndpoint = function (apiEndpoint) {
        this.apiEndpoint = apiEndpoint;
        return this;
    };
    CustomSmapiClientBuilder.prototype.withAuthEndpoint = function (authEndpoint) {
        this.authEndpoint = authEndpoint;
        return this;
    };
    CustomSmapiClientBuilder.prototype.withApiClient = function (apiClient) {
        this.apiClient = apiClient;
        return this;
    };
    /**
     * Funtion used to generate SkillManagementService instance.
     */
    CustomSmapiClientBuilder.prototype.client = function () {
        if (!this.apiEndpoint) {
            this.apiEndpoint = DEFAULT_API_ENDPOINT;
        }
        if (!this.apiClient) {
            this.apiClient = new DefaultApiClient();
        }
        if (this.refreshTokenConfig) {
            var apiConfiguration = {
                apiClient: this.apiClient,
                apiEndpoint: this.apiEndpoint,
                authorizationValue: null,
            };
            var authenticationConfiguration = {
                clientId: this.refreshTokenConfig.clientId,
                clientSecret: this.refreshTokenConfig.clientSecret,
                refreshToken: this.refreshTokenConfig.refreshToken,
                authEndpoint: this.authEndpoint,
            };
            return new ask_smapi_model_1.services.skillManagement.SkillManagementServiceClient(apiConfiguration, authenticationConfiguration, this.customUserAgent);
        }
        throw new Error('Please provide refreshToken Config to build smapi client');
    };
    return CustomSmapiClientBuilder;
}(StandardSmapiClientBuilder));
exports.CustomSmapiClientBuilder = CustomSmapiClientBuilder;
